compilers/concrete-compiler/compiler/lib/Dialect/FHE/Transforms/Optimizer.cpp

```diff
@@ -0,0 +1,112 @@
+// Part of the Concrete Compiler Project, under the BSD3 License with Zama
+// Exceptions. See
+// https://github.com/zama-ai/concrete-compiler-internal/blob/main/LICENSE.txt
+// for license information.
+
+#include <mlir/Transforms/GreedyPatternRewriteDriver.h>
+
+#include <concretelang/Dialect/FHE/Transforms/Optimizer/Optimizer.h>
+
+#include <llvm/Support/Debug.h>
+
+namespace mlir {
+namespace concretelang {
+
+struct OptimizerPartitionFrontierMaterializationPass
+    : public OptimizerPartitionFrontierMaterializationPassBase<
+          OptimizerPartitionFrontierMaterializationPass> {
+
+  OptimizerPartitionFrontierMaterializationPass(
+      const optimizer::CircuitSolution &solverSolution)
+      : solverSolution(solverSolution) {}
+
+  enum class OperationKind { PRODUCER, CONSUMER };
+
+  std::optional<uint64_t> getOid(mlir::Operation *op, OperationKind kind) {
+    if (mlir::IntegerAttr oidAttr =
+            op->getAttrOfType<mlir::IntegerAttr>("TFHE.OId")) {
+      return oidAttr.getInt();
+    } else if (mlir::DenseI32ArrayAttr oidArrayAttr =
+                   op->getAttrOfType<mlir::DenseI32ArrayAttr>("TFHE.OId")) {
+      assert(oidArrayAttr.size() > 0);
+
+      if (kind == OperationKind::CONSUMER) {
+        return oidArrayAttr[0];
+      } else {
+        return oidArrayAttr[oidArrayAttr.size() - 1];
```

Comment:
are you sure is always the case?

Reply:
From what I understand, a series of `OId`s indicates a sequence of operations in order from the operands to the results. Maybe @rudy-6-4 can confirm?

Reply:
Yes is that, but I'm not sure that every array of OIds have the last operation at the end of the array (need to check ConcreteOptimizer pass)

Reply:
This is what I understand e.g., from code handling the return value of `extractBitWithClearedLowerBits` in `FHEToTFHEScalar.cpp`. However, let's wait for @rudy-6-4 to confirm.

Reply:
I checked ConcreteOptimizer.cpp, looks like all TFHE.OId arrays follow the convention except the multiplication

```
    // Create the TFHE.OId attributes
    // The first elements of the vector are nodes for the encrypted
    // multiplication
    mlir::Builder builder(innerProductOp.getContext());
    mlir::SmallVector<int32_t, 8> operatorIndexes = {
        (int32_t)addNode.index,    (int32_t)lhsTluNode.index,
        (int32_t)subNode.index,    (int32_t)rhsCorrectionNode.index,
        (int32_t)rhsTluNode.index, (int32_t)resultNode.index,
    };
    if (lhsCorrectionNode.has_value()) {
      operatorIndexes.push_back(lhsCorrectionNode.value().index);
    }
    // The last element of the vector is the node for the addition
    operatorIndexes.push_back((int32_t)index[result].index);

    if (setOptimizerID)
      innerProductOp->setAttr("TFHE.OId",
                              builder.getDenseI32ArrayAttr(operatorIndexes));

    return index[result];
```

Reply:
Thanks for checking all the patterns. I added a special case to `OptimizerPartitionFrontierMaterializationPass::getOid()` for multiplications, so that it returns the OId at index 5.